Improve the tuple and unit trait docs

notriddle opened the pull request:
* Reduce duplicate impls; show only the `(T,)` and include a sentence saying that there exists ones up to twelve of them.
* Show `Copy` and `Clone`.
* Show auto traits like `Send` and `Sync`, and blanket impls like `Any`.

Here's the new version:

* <https://notriddle.com/notriddle-rustdoc-test/std/primitive.tuple.html>
* <https://notriddle.com/notriddle-rustdoc-test/std/primitive.unit.html>

notriddle committed:
docs: clean up trait docs for tuples

notriddle committed:
rustdoc: show auto/blanket docs for tuple and unit

notriddle committed:
docs: show Clone and Copy on () doc pages

rust-highfive commented:
Hey! It looks like you've submitted a new PR for the library teams!

If this PR contains changes to any `rust-lang/rust` public library APIs then please comment with `r? rust-lang/libs-api @rustbot label +T-libs-api -T-libs` to request review from a libs-api team reviewer. If you're unsure where your change falls no worries, just leave it as is and the reviewer will take a look and make a decision to forward on if necessary.

Examples of `T-libs-api` changes:

* Stabilizing library features
* Introducing insta-stable changes such as new implementations of existing stable traits on existing stable types
* Introducing new or changing existing unstable library APIs (excluding permanently unstable features / features without a tracking issue)
* Changing public documentation in ways that create new stability guarantees
* Changing observable runtime behavior of library APIs

rust-highfive commented:
r? @m-ou-se

(rust-highfive has picked a reviewer for you, use r? to override)

GuillaumeGomez commented:
I honestly have absolutely no idea. Providing this information unambigously and efficiently is gonna be the hard part I think.

cc @jsha @Manishearth

vacuus commented:
Perhaps
`impl<T0,...> Debug for (T0,...)`
`This trait is implemented for tuples up to twelve items long, where all items implement Debug. The above syntax is (currently) invalid.`

Manishearth commented:
I'm not sure either but yeah I do feel this is an area that could handle improvement.

I think `...` is a good way of doing it. We could also do a lil `*` on the `...` and have it show the "only implemented on tuples up till..." in a tooltip if we really want

jsha commented:
I think `impl Foo for (T, ...)` is good. The `(T, ...)` part should link to a subheading of the tuple page's top-doc that says:

> ### (T, ...)
> 
> In this documentation the shorthand (T, ...) is used to represent all tuples up to length twelve. When that is used, any trait bounds expressed on T applies to each field of the tuple independently. Note that this is a convenience notation to avoid repetitive documentation, not valid Rust syntax.

notriddle committed:
rustdoc: show tuple impls as `impl Trait for (T, ...)`

This commit adds a new unstable attribute, `#[doc(tuple_varadic)]`, that
shows a 1-tuple as `(T, ...)` instead of just `(T,)`, and links to a section
in the tuple primitive docs that talks about these.

notriddle committed:
rustdoc: fixed messed-up rustdoc auto trait impls

Before:

    impl<T, U> UnwindSafe for (T, ...) where
        T: UnwindSafe,
        U: UnwindSafe,

After:

    impl<T> UnwindSafe for (T, ...) where
        T: UnwindSafe,

notriddle committed:
Fix bootstrap attr

vacuus commented:
(I realize I'm being pedantic 😄 )

In the "Trait implementations" subheading:
> In this documentation the shorthand (T, ...) is used to represent all tuples up to length twelve.

But in the `impl Clone`/`impl Copy`:
> This trait is implemented on arbitrary-length tuples.

Instead, the subheading could say
> In this documentation the shorthand (T, ...) is used to represent all tuples up to some length, depending on the trait.

or outright state that it's a length of twelve except for `Clone`/`Copy`

jsha commented on compiler/rustc_span/src/symbol.rs:
```diff
@@ -1439,6 +1439,7 @@ symbols! {
         tuple,
         tuple_from_req,
         tuple_indexing,
+        tuple_varadic,
```
```suggestion
        tuple_variadic,
```

jsha commented on library/core/src/primitive_docs.rs:
```diff
@@ -895,24 +916,11 @@ mod prim_str {}
 ///
 /// # Trait implementations
 ///
-/// If every type inside a tuple implements one of the following traits, then a
-/// tuple itself also implements it.
-///
-/// * [`Clone`]
-/// * [`Copy`]
-/// * [`PartialEq`]
-/// * [`Eq`]
-/// * [`PartialOrd`]
-/// * [`Ord`]
-/// * [`Debug`]
-/// * [`Default`]
-/// * [`Hash`]
```
I like the explicitness of the old verbiage here. It may not strictly be needed with the new, improved display of trait implementations, but it's an alternate way to present the material and I think it's worth keeping.

jsha commented on library/std/src/primitive_docs.rs:
```diff
@@ -949,6 +957,31 @@ mod prim_str {}
 mod prim_tuple {}
 
+// Required to make auto trait impls render.
+// See /src/librustdoc/passes/collect_trait_impls.rs:collect_trait_impls
```
Elsewhere in the codebase, it looks like paths are expressed without this leading `/`.

```suggestion
// See src/librustdoc/passes/collect_trait_impls.rs:collect_trait_impls
```

jsha commented on src/librustdoc/html/format.rs:
```diff
@@ -1064,7 +1074,11 @@ impl clean::Impl {
+            if let clean::Type::Tuple(types) = &self.for_ &&
+                let [clean::Type::Generic(name)] = &types[..] &&
+                (self.kind.is_tuple_varadic() || self.kind.is_auto()) {
+                primitive_link_fragment(f, PrimitiveType::Tuple, &format!("({name}, ...)"), "#trait-implementations-1", cx)?;
```
Just a note to say that we are likely to eventually scope anchor fragments, such that this anchor would change, and the internal link would break. Here's a possible way to catch that: Write a test case that contains a `# Trait Implementations` heading, and verify that it contains this anchor. The test case could then have a comment indicating what needs to be updated.

jsha reviewed (changes requested):
The display and the text are looking good to me. Just to check: are the demo links up to date?

A note on consistency: the current demo has:

```
impl<T: Clone> Clone for (T, ...)
impl<T11> Debug for (T11, ...)
impl<L> Default for (L, ...)
impl<A> Hash for (A, ...)
```

I haven't fully grokked the macros yet, so it's not clear to me where the `T`, `T11`, `L`, and `A` are coming from. It would be nice for them to be consistently `T` unless the other type variable names mean something.

notriddle committed:
Fix incorrectly spelled "variadic"

notriddle committed:
Use relative path for addressing things in rust-lang/rust

Co-authored-by: Jacob Hoffman-Andrews <github@hoffman-andrews.com>

notriddle committed:
Re-add explicit list of traits to tuple docs, with limit notes

notriddle committed:
Add test case for #trait-implementations-1 link

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6844480986?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/c1487550ca75c2bcd4cb4da038ef1d18d86a857a/checks/6844480986/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
diff of stderr:

11   --> $DIR/feature-gate-rustdoc_internals.rs:7:1
12    |
13 LL | #[doc(tuple_variadic)]
+    | ^^^^^^^^^^^^^^^^^^^^^^
15    |
16    = note: see issue #90418 <https://github.com/rust-lang/rust/issues/90418> for more information
17    = help: add `#![feature(rustdoc_internals)]` to the crate attributes to enable
---
To only update this specific test, also pass `--test-args feature-gates/feature-gate-rustdoc_internals.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/feature-gates/feature-gate-rustdoc_internals.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/feature-gate-rustdoc_internals" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/feature-gate-rustdoc_internals/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0658]: `#[doc(keyword)]` is meant for internal use only
   |
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
LL | #[doc(keyword = "match")] //~ ERROR: `#[doc(keyword)]` is meant for internal use only
   |
   = note: see issue #90418 <https://github.com/rust-lang/rust/issues/90418> for more information
   = help: add `#![feature(rustdoc_internals)]` to the crate attributes to enable


error[E0658]: `#[doc(tuple_variadic)]` is meant for internal use only
   |
   |
LL | #[doc(tuple_variadic)]  //~ ERROR: `#[doc(tuple_variadic)]` is meant for internal use only
   |
   = note: see issue #90418 <https://github.com/rust-lang/rust/issues/90418> for more information
   = help: add `#![feature(rustdoc_internals)]` to the crate attributes to enable

```

</details>

notriddle committed:
Fix broken test case

notriddle committed:
docs: make all the variadic impls use `(T, ...)` exactly

GuillaumeGomez commented:
I like it like that. For `clone`, it only says:

> This trait is implemented on arbitrary-length tuples.

Shouldn't it be `up to twelve elements` there as well?

GuillaumeGomez commented:
Oh nice! TIL.

jsha commented on src/librustdoc/clean/types.rs:
```diff
@@ -1839,10 +1839,10 @@ impl PrimitiveType {
                 // Either manually update this arrayvec at this point
                 // or start with a more complex refactoring.
-                Tuple => [TupleSimplifiedType(2), TupleSimplifiedType(3)].into(),
+                Tuple => [TupleSimplifiedType(1), TupleSimplifiedType(2), TupleSimplifiedType(3)].into(),
```
I've poked around this part of the code in an IDE and I can't figure out what it does. Would you mind briefly explaining it to me, along with why it needed to change?

jsha replied:
Thanks!

jsha commented on compiler/rustc_passes/src/check_attr.rs:
```diff
@@ -804,6 +804,37 @@ impl CheckAttrVisitor<'_> {
+                        .sess
+                        .struct_span_err(
+                            meta.span(),
+                            "`#[doc(tuple_variadic)]` can only be used on unary tuples",
```
Maybe I'm alone on this but I found this error confusing, since tuple_variadic is intended specifically for implementations that cover tuples with arity greater than one.

How about: `#[doc(tuple_variadic)] can only be used on the first of a set of tuple trait implementations with varying arity` or similar?

jsha reviewed (changes requested):
The generated docs look great, and the code looks great; down to nits and documentation requests now.

Once those are resolved can I r= as a member of the rustdoc team or do we also need an approval from the libs team?

notriddle committed:
Update library/std/src/primitive_docs.rs

Co-authored-by: Jacob Hoffman-Andrews <github@hoffman-andrews.com>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6847259214?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/80b201da7d2b53024d595403809906b9205809e2/checks/6847259214/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 8.65s
tidy check
tidy: Skipping binary file check, read-only filesystem
tidy error: library/{core_name} and library/{std_name} have different contents
* 630 error codes
* highest error code: E0788
Found 505 error codes
Found 0 error(s) in error codes
```

</details>

notriddle committed:
Update library/core/src/primitive_docs.rs

notriddle committed:
Add docs to `maybe_tuple_doc!`

GuillaumeGomez commented:
> Once those are resolved can I r= as a member of the rustdoc team or do we also need an approval from the libs team?

Since we're making a certain amount of changes into the public std API display, we might as well ask them.

cc @rust-lang/libs-api

BurntSushi commented:
I _think_ the rendering LGTM and also looks like an improvement. I'm not 100% certain, but I don't have any better ideas.

GuillaumeGomez commented:
We can always improve it later on if someone comes up with a better idea.

notriddle committed:
rustdoc: change error message for invalid `#[doc(tuple_variadic)]`

Add test case.

jsha commented on compiler/rustc_passes/src/check_attr.rs:
```diff
@@ -815,7 +815,7 @@ impl CheckAttrVisitor<'_> {
                         .struct_span_err(
                             meta.span(),
-                            "`#[doc(tuple_variadic)]` can only be used on unary tuples",
+                            "`#[doc(tuple_variadic)]` must used on the first of a set of tuple trait impls with varying arity",
```
```suggestion
                            "`#[doc(tuple_variadic)]` must be used on the first of a set of tuple trait impls with varying arity",
```

jsha commented on library/core/src/tuple.rs:
```diff
@@ -19,75 +19,104 @@ macro_rules! tuple_impls {
     }
 }
 
+macro_rules! maybe_tuple_doc {
```
This is great. Succinct but clear.

notriddle committed:
rustdoc: add missing article

Co-authored-by: Jacob Hoffman-Andrews <github@hoffman-andrews.com>

jsha commented:
Looks great! r=me; maybe give another day for any last feedback from libs team?

m-ou-se commented:
> `impl<Tᵢ: Copy> Copy for (T₁, T₂, …, Tₙ)` too?

Hm, that might result in people wondering what the `i` represents. (And adding more notation with `∀` and `i ∈ {1, 2, …, n}` or something like that probably goes a bit too far. ^^)

jsha commented:
> Maybe, to avoid being overly verbose, it could use subscripts in the tuple, but not elsewhere?
> `impl<T: Copy> Copy for (T₁, T₂, …, Tₙ)`

I like this version. And the detail that T's bound must apply to each Tₙ individually can be explained in the linked paragraph in the documentation.

notriddle committed:
rustdoc: change "variadic tuple" notation to look less like real syntax

GuillaumeGomez commented:
Looks almost all good to me. Can you add a test for the notation as well please?

notriddle committed:
rustdoc: add test case for "variadic tuple" search notation

GuillaumeGomez commented:
Looks good to me, thanks!

GuillaumeGomez commented:
Let's go then. Thanks everyone for the feedback!

@bors r=jsha,GuillaumeGomez

bors commented:
:pushpin: Commit f1d24beb6805e4d16d472964518c576421f8c009 has been approved by `jsha,GuillaumeGomez`

<!-- @bors r=jsha,GuillaumeGomez f1d24beb6805e4d16d472964518c576421f8c009 -->
<!-- homu: {"type":"Approved","sha":"f1d24beb6805e4d16d472964518c576421f8c009","approver":"jsha,GuillaumeGomez"} -->

bors commented:
:hourglass: Testing commit f1d24beb6805e4d16d472964518c576421f8c009 with merge 6ec3993ef4a4eb72bc20477fe9a4d92acd53f2c6...
<!-- homu: {"type":"BuildStarted","head_sha":"f1d24beb6805e4d16d472964518c576421f8c009","merge_sha":"6ec3993ef4a4eb72bc20477fe9a4d92acd53f2c6"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6919714560?check_suite_focus=true)
Approved by: jsha,GuillaumeGomez
Pushing 6ec3993ef4a4eb72bc20477fe9a4d92acd53f2c6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jsha,GuillaumeGomez","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6919714560?check_suite_focus=true"},"merge_sha":"6ec3993ef4a4eb72bc20477fe9a4d92acd53f2c6"} -->

WaffleLapkin commented on library/core/src/tuple.rs:
```diff
@@ -125,4 +156,4 @@ macro_rules! last_type {
 }
 
-tuple_impls!(A B C D E F G H I J K L);
+tuple_impls!(E D C B A Z Y X W V U T);
```
Is there a particular reason for this change?

WaffleLapkin replied:
Wouldn't `s/L/T` suffice then? The letters seem random...

jsha commented:
I think `impl Foo for (T, ...)` is good. The `(T, ...)` part should link to
a subheading of the tuple page's top-doc that say, approximately, "In this
documentation (T, ...) is used to represent all tuples up to length twelve.
In such constructions, any trait bounds expressed on T apply to all fields
of the tuple. Not that this is a shorthand, not valid Rust syntax."
